pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java

```diff
@@ -152,39 +202,82 @@ private TransferableBlock buildJoinedDataBlock(TransferableBlock leftBlock)
     if (leftBlock.isErrorBlock()) {
       _upstreamErrorBlock = leftBlock;
       return _upstreamErrorBlock;
-    } else if (TransferableBlockUtils.isNoOpBlock(leftBlock) || TransferableBlockUtils.isEndOfStream(leftBlock)) {
+    }
+    if (leftBlock.isNoOpBlock() || (leftBlock.isSuccessfulEndOfStreamBlock() && !needUnmatchedRows())) {
       return leftBlock;
     }
+    if (leftBlock.isSuccessfulEndOfStreamBlock()) {
```

Comment:
nit: suggest putting the logic of the end of stream handling into a separate function for better readability.

Reply:
I added a TODO to follow up.